sasl.go

```diff
@@ -162,66 +162,65 @@ type saslXOAUTH2Handler struct {
 	errorResponse        []byte // https://developers.google.com/gmail/imap/xoauth2-protocol#error_response
 }
 
-func (s saslXOAUTH2Handler) init() stateFunc {
+func (s saslXOAUTH2Handler) init() (stateFunc, error) {
 	originalPeerMaxFrameSize := s.conn.PeerMaxFrameSize
 	if s.maxFrameSizeOverride > s.conn.PeerMaxFrameSize {
 		s.conn.PeerMaxFrameSize = s.maxFrameSizeOverride
 	}
-	s.conn.err = s.conn.writeFrame(frames.Frame{
+	err := s.conn.writeFrame(frames.Frame{
 		Type: frameTypeSASL,
 		Body: &frames.SASLInit{
 			Mechanism:       saslMechanismXOAUTH2,
 			InitialResponse: s.response,
 		},
 	})
 	s.conn.PeerMaxFrameSize = originalPeerMaxFrameSize
-	if s.conn.err != nil {
-		return nil
+	if err != nil {
+		return nil, err
 	}
 
-	return s.step
+	return s.step, nil
 }
 
-func (s saslXOAUTH2Handler) step() stateFunc {
+func (s saslXOAUTH2Handler) step() (stateFunc, error) {
 	// read challenge or outcome frame
 	fr, err := s.conn.readFrame()
 	if err != nil {
-		s.conn.err = err
-		return nil
+		return nil, err
```

Comment:
Trying to understand this particular part.

It seems "correct" to set s.conn.err here, since the connection is in a failed state. You've switched this code over to be non-modifying though, so I think you have a plan. 

Where do you see this code going? I can see a few positives here, but not sure if we're thinking the same things. For instance:

- We can consider retrying on some of these busted states, thus bringing the connection from an error state into a correct state. 
- It'd be better to centralize